MaxText/standalone_checkpointer.py

```diff
@@ -0,0 +1,99 @@
+"""
+ Copyright 2023 Google LLC
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
+# pylint: disable=g-bad-todo, abstract-method, consider-using-with, ungrouped-imports
+"""Training loop and Decoding of the model."""
+
+# Calling jax.device_count here prevents a "TPU platform already registered" error.
+# See github.com/google/maxtext/issues/20 for more
+
+import datetime
+import os
+import sys
+
+from typing import Sequence
+from absl import app
+from flax import linen as nn
+from flax.linen import partitioning as nn_partitioning
+import jax
+import numpy as np
+import optax
+from tensorboardX import SummaryWriter
+
+import checkpointing
+import max_utils
+import maxtext_utils
+import max_logging
+import pyconfig
+from train import validate_train_config, calculate_training_tflops, get_first_step, load_next_batch, setup_train_loop
+
+from input_pipeline import create_data_iterator_with_tokenizer
+from layers import models
+
+import jax.numpy as jnp
+from jax import random
+from jax.sharding import Mesh
+
+from jax.experimental.compilation_cache import compilation_cache as cc
+
+Transformer = models.Transformer
+
+def checkpoint_loop(config, state=None):
+  """Main Checkpointing loop.
+  Saves checkpoints.
+  Args:
+    config:
+    state:
+    ckpt_path:
+  Returns:
+  """
+  init_rng = random.PRNGKey(config.init_weights_seed)
+  ( writer, checkpoint_manager, _, _, _, _, _, state) = setup_train_loop(config, init_rng, log_checkpoint_restore_time=True)
+  start_step = get_first_step(state) # this is the start_step for training
+
+  for step in np.arange(start_step, config.steps):
+    start_time = datetime.datetime.now()
+    if checkpoint_manager is not None:
+      if checkpoint_manager.save(step, state):
+        max_logging.log(f"saved a checkpoint at step {step}")
+        checkpoint_manager.wait_until_finished()
```

Comment:
@rwitten and @gobbleturk - need some help here. I was hoping that this function is the Orbax equivalent of jax.block_until_ready(), but I am not quite sure. I see multiple logs , for each of the hosts.

Reply:
wait_until_finished() shoudl block (and DEFINITELY should block if async_checkpointing=false). Is there a problem?

Reply:
I was hoping save() to be blocking . I found out that wait_until_finished() blocks in case of async checkpoint. 
Even with jax.experimental.multihost_utils.sync_global_devices("dummy") before the start of checkpointing, I see differences in the order of microseconds between the hosts. 

I decided not to be nit-picky about this at the moment.

Reply:
Added now!